### PR TITLE
Post merge review updates for CommandHandler updates

### PR DESCRIPTION
**Pull request by andy31415**
This addresses post-merge review comments in #33620.

### Changes

- Move `TryAddResponseData` into cpp file instead of h
- move EncoderToTLV into `DataModel` (path and namespace) and rename `DataModelEncoderToTLV` to `EncodableType` (so that DataModel:: does not stutter)
- removed `inline` hint for CommandHandler
- Removed the `ViaEncoder` suffix to methods as method arguments are different
- Various typo fixes in comments and updated some comment descriptions

### CODE REVIEW

**Review by tehampson — APPROVED** [review]

File: src/app/data-model/EncoderToTLV.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/*
+ *    Copyright (c) 2020-2021 Project CHIP Authors
```
Comment (tehampson) [inline src/app/data-model/EncoderToTLV.h:2]: nit, 2024

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/CommandHandler.cpp [hunk 2]
```diff
@@ -113,6 +113,17 @@ Status CommandHandler::OnInvokeCommandRequest(CommandHandlerExchangeInterface &
     return status;
 }
 
+CHIP_ERROR CommandHandler::TryAddResponseData(const ConcreteCommandPath & path, CommandId commandId,
+                                              DataModel::EncoderToTLV & encoder)
+{
+    ConcreteCommandPath responseCommandPath = { path.mEndpointId, path.mClusterId, commandId };
+    ReturnErrorOnFailure(TryAddResponseDataPreEncode(path, responseCommandPath));
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.cpp:120]: The split between TryAddResponseData and TryAddResponseDataPreEncode was an artificial distinction due to the templating: TryAddResponseDataPreEncode was non-templated and shared across instantiations.

Do we still want/need this split?  Is there a reason to have TryAddResponseDataPreEncode be a separate function?  And does it need to be inlined?  The documentation for this function no longer makes sense in the new setup, by the way; the whioe "non-templated... " bit is not relevant.

File: src/app/CommandHandler.h [hunk 3]
```diff
@@ -374,13 +347,12 @@ class CommandHandler
      * Most applications are likely to use `AddResponseData` as a more convenient
      * one-call that auto-sets command ID and creates the underlying encoders.
      */
-    CHIP_ERROR AddResponseDataViaEncoder(const ConcreteCommandPath & aRequestCommandPath, CommandId commandId,
-                                         EncoderToTLV & encoder)
+    CHIP_ERROR AddResponseData(const ConcreteCommandPath & aRequestCommandPath, CommandId commandId,
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:350]:
```suggestion
    CHIP_ERROR AddResponseData(const ConcreteCommandPath & aRequestCommandPath, CommandId aResponseCommandId,
```
Two things here:

1. Follow file style for argument naming.
2. Make it clear that this is the command ID of the response command.  It took me a while to understand why we need a command ID separate from the one in aRequestCommandPath.

File: src/app/CommandHandler.h [hunk 4]
```diff
@@ -374,13 +347,12 @@ class CommandHandler
...
+                               DataModel::EncoderToTLV & encoder)
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:351]:
```suggestion
                               DataModel::EncoderToTLV & aEncoder)
```
But note comments below about EncoderToTLV naming.

File: src/app/CommandHandler.h [hunk 5]
```diff
@@ -410,9 +382,9 @@ class CommandHandler
      * The encoder would generally encode a ClusterName::Commands::CommandName::Type with
      * the corresponding `GetCommandId` call.
      */
-    void AddResponseViaEncoder(const ConcreteCommandPath & aRequestCommandPath, CommandId commandId, EncoderToTLV & encoder)
+    void AddResponse(const ConcreteCommandPath & aRequestCommandPath, CommandId commandId, DataModel::EncoderToTLV & encoder)
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:385]:
```suggestion
    void AddResponse(const ConcreteCommandPath & aRequestCommandPath, CommandId aResponseCommandId, DataModel::EncoderToTLV & aEncoder)
```

File: src/app/CommandHandler.h [hunk 6]
```diff
@@ -695,20 +667,11 @@ class CommandHandler
...
-                                            EncoderToTLV & encoder)
-    {
-        ConcreteCommandPath responseCommandPath = { aRequestCommandPath.mEndpointId, aRequestCommandPath.mClusterId, commandId };
-        ReturnErrorOnFailure(TryAddResponseDataPreEncode(aRequestCommandPath, responseCommandPath));
-        TLV::TLVWriter * writer = GetCommandDataIBTLVWriter();
-        VerifyOrReturnError(writer != nullptr, CHIP_ERROR_INCORRECT_STATE);
-        ReturnErrorOnFailure(encoder.Encode(*writer, TLV::ContextTag(CommandDataIB::Tag::kFields)));
-        return FinishCommand(/* aEndDataStruct = */ false);
-    }
+    CHIP_ERROR TryAddResponseData(const ConcreteCommandPath & path, CommandId commandId, DataModel::EncoderToTLV & encoder);
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:674]: This is gong the wrong direction in terms of argument naming.  File style is `aSomething`.  And note above comments about command ID naming.

File: src/app/data-model/EncoderToTLV.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+#include <lib/core/CHIPError.h>
+#include <lib/core/TLV.h>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/// Defines an abstract class of something that can be encoded
+/// into a TLV with a given data tag
+class EncoderToTLV
```
Comment (bzbarsky-apple) [inline src/app/data-model/EncoderToTLV.h:30]: Seems like EncodableToTLV would be a better name for this class.  This thing is not an encoder; it is encodable....

File: src/app/data-model/EncoderToTLV.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    virtual CHIP_ERROR Encode(TLV::TLVWriter & writer, TLV::Tag tag) = 0;
+};
+
+/// An `EncoderToTLV` that uses `DataModel::Encode` to encode things.
+///
+/// Generally useful to encode things like <ClusterName>::Commands::<CommandName>::Type
+/// structures.
+template <typename T>
+class ObjectEncoderToTLV : public EncoderToTLV
```
Comment (bzbarsky-apple) [inline src/app/data-model/EncoderToTLV.h:43]: How about DataModelTypeEncodableToTLV?  Or ClusterObjectEncodableToTLV?

It's really unclear what "Object" on its own is doing there.  What sort of object?  Note that "value" here can totally be an integer or something....

Comment (andy31415) [reply src/app/data-model/EncoderToTLV.h]: The namespace was frustrating: `DataModel::DataModel....` stutters.

I picked Object for now to not stutter and this is already inside DataModel.

Comment (andy31415) [reply src/app/data-model/EncoderToTLV.h]: Picked `DirectEncodableToTLV` ... at least not `Object` anymore.

File: src/app/data-model/EncodableToTLV.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+namespace DataModel {
+
+/// Defines an abstract class of something that can be encoded
+/// into a TLV with a given data tag
+class EncodableToTLV
+{
+public:
+    virtual ~EncodableToTLV() = default;
+
+    virtual CHIP_ERROR EncodeTo(TLV::TLVWriter & writer, TLV::Tag tag) = 0;
```
Comment (bzbarsky-apple) [inline src/app/data-model/EncodableToTLV.h:35]: Should be const?

File: src/app/data-model/EncodableToTLV.h [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+/// An `EncodableToTLV` that uses `DataModel::Encode` to encode things in one call.
+///
+/// Applicable to any type for which `chip::app::DataModel::Encode` works. In
+/// particular, types like <ClusterName>::Commands::<CommandName>::Type
+/// can be used as a type here.
+template <typename T>
+class DirectEncodableToTLV : public EncodableToTLV
```
Comment (bzbarsky-apple) [inline src/app/data-model/EncodableToTLV.h:44]: How about:
```suggestion
template <typename DataModelType>
class EncodableDataModelType : public EncodableToTLV
```

File: src/app/data-model/EncoderToTLV.h (hunk 8 above)
Comment (andy31415) [reply src/app/data-model/EncoderToTLV.h]: Updated again to `EncodableType`

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**